### PR TITLE
Fix-EZP-27373: Error when edit empty not required date

### PR DESCRIPTION
**Pull request by kmadejski**
JIRA issue: [https://jira.ez.no/browse/EZP-27373](https://jira.ez.no/browse/EZP-27373)

### Description
If `Date` FieldType is used and **Searchable** setting is checked and also default value is set to **Empty** then it is impossible to save content without setting the value of this field. 
API returns an error as described in issue.

### CODE REVIEW

File: eZ/Publish/Core/FieldType/Date/SearchField.php [hunk 1]
```diff
@@ -29,12 +29,17 @@ class SearchField implements Indexable
      */
     public function getIndexData(Field $field, FieldDefinition $fieldDefinition)
     {
-        $dateTime = new DateTime("@{$field->value->data['timestamp']}");
+        if ($field->value->data['timestamp']) {
```
Comment (kmadejski) [inline eZ/Publish/Core/FieldType/Date/SearchField.php]: @alongosz You are definitely right.

**Comment by alongosz** [comment]
@kmadejski I took a closer look at PRs and it looks like this is related to #1967 and #1992 as a follow-up on #1942.
This seems to affect `DateAndTime` field, so same fix for `SearchField` there is required.

@gggeek You've mentioned in [#1942](https://github.com/ezsystems/ezpublish-kernel/pull/1942#issuecomment-295011095) that you'd prefer here cast to int.
Casting `null` to `int` gives us 0, so it means 1 Jan 1970. However if `null` gets passed here, it rather means that `SearchField` value should be `null`, thus not indexed (IMHO). 
Is this ok with you?

**Comment by andrerom** [comment]
@kmadejski See the other PR's above, there are some more cases to cover.

**Comment by kmadejski** [comment]
@alongosz @andrerom thank you, I didn't find it before, I will investigate it more.

**Comment by alongosz** [comment]
Note: one of the cases involves `Value::fromTimestamp` method. I wouldn't be so sure about it, because this method is designed to get `int` and only `int`. If it gets something else, like `null` it should throw `InvalidArgument` exception, which it already does.

**Comment by kmadejski** [comment]
@alongosz, @andrerom  Correct me if I'm wrong, but it seems `Value::fromTimestamp` is invoked always with `int` values. See: 
https://github.com/ezsystems/ezpublish-kernel/blob/master/eZ/Publish/Core/FieldType/Date/Type.php#L91-L93
and as you mentioned before, if we cast `null` to `int` we will get 0, so I think that this code should work well:
https://github.com/ezsystems/ezpublish-kernel/blob/master/eZ/Publish/Core/FieldType/Date/Type.php#L153

The same situation in case of `DateAndTime` FieldType, see:
https://github.com/ezsystems/ezpublish-kernel/blob/master/eZ/Publish/Core/FieldType/DateAndTime/Type.php#L101-L103
https://github.com/ezsystems/ezpublish-kernel/blob/master/eZ/Publish/Core/FieldType/DateAndTime/Type.php#L170

As well, we can go forward with @yannickroger idea proposed in https://github.com/ezsystems/ezpublish-kernel/pull/1967/, I mean to cast all given values to `int` directly in `SearchField::getIndexData` and return it always as `int` even it is 0.

**Comment by andrerom** [comment]
@kmadejski you are right, it's documented to only support int.

**Comment by kmadejski** [comment]
@andrerom Thanks, I will prepare fix also for `DateAndTime/SearchField.php`, it seems to be the same issue there.

**Comment by andrerom** [comment]
@kmadejski Ok, then also check Time, and make sure the originally issue fixed in #1942 is solved for all of them.

**Comment by kmadejski** [comment]
@andrerom fixed also for `Time` and `DateAndTime` FieldTypes. I did manual tests and solution looks working as expected.

**Comment by kmadejski** [comment]
@alongosz I changed validation in `Time` regarding your advice. Thank you.

File: eZ/Publish/Core/FieldType/DateAndTime/SearchField.php [hunk 2]
```diff
@@ -28,10 +29,17 @@ class SearchField implements Indexable
      */
     public function getIndexData(Field $field, FieldDefinition $fieldDefinition)
     {
+        if (isset($field->value->data['timestamp']) && $field->value->data['timestamp'] !== '') {
```
Comment (alongosz) [inline eZ/Publish/Core/FieldType/DateAndTime/SearchField.php]: `if ($field->value->data !== null)` is enough because DateAndTime Converter [ensures the same thing as above](https://github.com/ezsystems/ezpublish-kernel/blob/v6.7.3/eZ/Publish/Core/Persistence/Legacy/Content/FieldValue/Converter/DateAndTimeConverter.php#L57-L68).

File: eZ/Publish/Core/FieldType/Time/SearchField.php [hunk 3]
```diff
@@ -28,10 +29,17 @@ class SearchField implements Indexable
      */
     public function getIndexData(Field $field, FieldDefinition $fieldDefinition)
     {
+        if ($field->value->data !== null && $field->value->data !== '') {
```
Comment (alongosz) [inline eZ/Publish/Core/FieldType/Time/SearchField.php]: `if ($field->value->data !== null)` is enough because `TimeConverter` also [ensures](https://github.com/ezsystems/ezpublish-kernel/blob/v6.7.3/eZ/Publish/Core/Persistence/Legacy/Content/FieldValue/Converter/TimeConverter.php#L55-L63) `data` remains `null` if not set.

**Comment by alongosz** [comment]
I've also found another bug related to [this line](https://github.com/ezsystems/ezpublish-kernel/blob/master/eZ/Publish/Core/Persistence/Legacy/Content/FieldValue/Converter/DateConverter.php#L59).
Basically if you set your DateTime to 1st Jan 1970, which is `0`, reloading content will not set `data` property. Removing ` || $value->dataInt == 0` seems to fix it.

Not sure if this can be included here or should be a separate PR. @andrerom?

Test cases for `eZ\Publish\API\Repository\Tests\FieldType\DateIntegrationTest`:
```php
    /**
     * @return \eZ\Publish\API\Repository\Values\Content\VersionInfo
     */
    public function testCreateContentWithZeroDate()
    {
        $fieldValue = DateValue::fromTimestamp(0);
        $content = $this->createContent($fieldValue);
        $loadedContent = $this->getRepository(false)->getContentService()->loadContentByContentInfo($content->contentInfo);
        self::assertEquals($content, $loadedContent);

        return $loadedContent->versionInfo;
    }

    /**
     * @depends testPublishContentWithZeroDate
     *
     * @param \eZ\Publish\API\Repository\Values\Content\VersionInfo $versionInfo
     */
    public function testPublishContentWithZeroDate(VersionInfo $versionInfo)
    {
        $this->getRepository(false)->getContentService()->publishVersion($versionInfo);
    }
```

File: eZ/Publish/Core/FieldType/DateAndTime/SearchField.php (hunk 2 above)
Comment (andrerom) [inline eZ/Publish/Core/FieldType/DateAndTime/SearchField.php]: And we are sure this only comes from storage and not user ?

**Comment by kmadejski** [comment]
@alongosz don't worry, your advice are always very helpful, thank you! I'll fix it in few minutes.

**Comment by andrerom** [comment]
> Not sure if this can be included here or should be a separate PR. @andrerom?

lets make that separate, this is already covering to many things (regression and second look at the original issue).

File: eZ/Publish/Core/FieldType/DateAndTime/SearchField.php (hunk 2 above)
Comment (alongosz) [inline eZ/Publish/Core/FieldType/DateAndTime/SearchField.php]: > And we are sure this only comes from storage and not user (via input validation and such)?

Yes. `getIndexData` is called by `eZ\Publish\SPI\Search\Handler::indexContent` which always gets its content from `eZ\Publish\SPI\Persistence\Content\Handler::load`

**Comment by alongosz** [comment]
And we've uncovered two separate issues...
- `InputParserPassTest::testProcess` [fails](https://travis-ci.org/ezsystems/ezpublish-kernel/jobs/237480102#L704) for Symfony 3.3, because service ids are case sensitive (or rather will be in the future). To fix this in a BC way let's change `UnitTest` to lowercase in this method.
- Empty MapLocation for Elasticsearch is incorrectly mapped, so [fails](https://travis-ci.org/ezsystems/ezpublish-kernel/jobs/237480098#L705) (hurrah, we've found actual bug!)
Looking at ES logs it seems that the method
`eZ\Publish\Core\Search\Elasticsearch\Content\FieldValueMapper\GeoLocationMapper::map` should have an extra check, like:
```php
        if (!isset($field->value['latitude'], $field->value['longitude'])) {
            return null;
        }
```
These probably need to be separate PRs.

**Comment by andrerom** [comment]
Looks good now, but one control question:
> Same solution was provided also in Time and DateAndTime FieldTypes.

Was Time and DateAndTime actually affected by the timezone issue solved in #1942? Or could they just have been kept as they where _(as in just passing on the timezone, if there is one. aka only fix would be to make sure to check that data is set in DateAndTime)_.

**Comment by kmadejski** [comment]
@andrerom `DateAndTime` wasn't affected with https://github.com/ezsystems/ezpublish-kernel/pull/1942 because there always was @ in `Value::fromTimestamp`, but @ was missing in `SearchField::getIndexData` so it can cause a similar error like this one related to this PR or returns the wrong date to the search index.

In `Time` problem described in https://github.com/ezsystems/ezpublish-kernel/pull/1942 does not exist because timestamp is passed in method `DateTime::setTimestamp` which is timezone aware.
There is also @ missing in `SearchField::getIndexData`, but in this case, I'm not sure if it is really needed. @alongosz what do you think?

**Comment by andrerom** [comment]
> Should I implement something from things mentioned by @alongosz here: #1996 (comment)?

Hmm, well you can, the second one is caused by tour new test, as this is depending on each other I suggest we end up with three commits that can be merged normally (not squashed):
- commit to add testing of empty fields _(several fields should fail on this commit alone it seems by the look of this PR and it's travis failures)_
- commit to fix date/time/dateAndTime where needed
- commit to fix GeoLocationMapper::map


+ one separate PR for Symfony 3.3, @alongosz can you handle that one? _(6.7 and up to unblock PR's)_

**Comment by alongosz** [comment]
> Was Time and DateAndTime actually affected by the timezone issue solved in #1942?

It wasn't AFAIR.
If we look at the original code for [`DateAndTime\SearchField`](https://github.com/ezsystems/ezpublish-kernel/blob/v6.7.3/eZ/Publish/Core/FieldType/DateAndTime/SearchField.php#L34) and [`Time\SearchField`](https://github.com/ezsystems/ezpublish-kernel/blob/v6.7.3/eZ/Publish/Core/FieldType/Time/SearchField.php#L34), integers are indexed there *which makes no sense* for search engine. Code itself will work, as in PHP `null['whatever']` silently results in `null`... (@Nattfarinn remindend me of this today :-))

Anyway, I thought this would be a good improvement, but it's probably a little bit out of scope of PR. One could also argue that it's a BC break, but it's hard for me to imagine someone actually relying on unix timestamps in search index ;)

**Comment by alongosz** [comment]
> one separate PR for Symfony 3.3, @alongosz can you handle that one? (6.7 and up to unblock PR's)

#1998

**Comment by kmadejski** [comment]
@andrerom I discussed it with @alongosz and it seems to be unnecessary in FieldTypes other than `Date` so right now I kept changes only there. Added commit with test and also fix for `GeoLocationMapper`. Is it okay?

**Comment by andrerom** [comment]
@kmadejski Almost. 1. you have 6 commits and not 3 :) and 2. while at it rebase on 6.7 which now has fix for symfony 3.3.

**Comment by slaci** [comment]
Is that possible: this `getIndexData` is called for *non* `searchable` fields too on publish?

I have an `ezdate` field where searchable is 0 (I can see it in ezcontentclass_attribute table too its really 0), but publish fails with the invalidargument exception, cause the timestamp is null.

This only happens with solr, not with legacy search.

So currently ezdate cannot be optional :\

**Comment by kmadejski** [comment]
@slaci I checked it right now, content can be published without errors when is *not* `Searchable` no matter which search engine is used. I don't know what caused your problem. Anyway, this solution might be helpful also in the case which you described.
